### PR TITLE
[Men-33] - Split the storage node in multiple sub-nodes

### PR DESCRIPTION
**Pull request by dalloriam**

### CODE REVIEW

File: bin/menmosd/src/menmosd/node/service/admin.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+#[async_trait]
+impl interface::NodeAdminController for NodeAdminService {
+    async fn register_storage_node(&self, def: StorageNodeInfo) -> Result<StorageNodeResponseData> {
+        let rebuild_requested = {
+            let rebuild_queue_guard = self.rebuild_queue.lock().map_err(|e| anyhow!("{}", e))?;
+
+            if let Some(storage_node) = rebuild_queue_guard.last() {
+                storage_node.id == def.id
+            } else {
+                false
```
Comment (gabrielmethot) [inline bin/menmosd/src/menmosd/node/service/admin.rs:40]: Why is that necessary?

File: bin/menmosd/src/menmosd/node/service/indexer.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            if info.owner != username {
+                return Ok(None);
+            }
+        }
+
+        Ok(blob_info_maybe)
+    }
+
+    async fn get_blob_storage_node(&self, blob_id: &str) -> Result<Option<StorageNodeInfo>> {
+        // These next two lines could technically be in one line, but since its an async
```
Comment (gabrielmethot) [inline bin/menmosd/src/menmosd/node/service/indexer.rs:72]: Very nice

File: bin/menmosd/src/menmosd/node/service/query.rs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        if resulting_bitvector.count_ones() == 0 {
+            // No pending move requests.
+            return Ok(move_requests);
+        }
+
+        for doc_idx in resulting_bitvector.iter_ones() {
+            let blob_id = self
+                .documents
+                .lookup(doc_idx as u32)?
+                .ok_or_else(|| anyhow!("missing doucment ID for index '{}'"))?;
```
Comment (gabrielmethot) [inline bin/menmosd/src/menmosd/node/service/query.rs:131]: You made a typo: "doucment" -> "document"

File: bin/menmosd/src/menmosd/node/service/query.rs [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        let tag_list = match tags.as_ref() {
+            Some(tag_filters) => {
+                let mut hsh = HashMap::with_capacity(tag_filters.len());
+                for tag in tag_filters {
+                    hsh.insert(
+                        tag.clone(),
+                        (self.metadata.load_tag(tag)? & user_mask.clone()).count_ones(),
+                    );
+                }
+                hsh
```
Comment (gabrielmethot) [inline bin/menmosd/src/menmosd/node/service/query.rs:163]: Why is that necessary?

File: bin/menmosd/src/menmosd/node/test/mock.rs [hunk 5]
```diff
@@ -21,14 +38,56 @@ fn tag_to_kv(tag: &str) -> Result<(&str, &str)> {
...
+        Ok(())
+    }
+
+    fn has_user(&self, username: &str) -> Result<bool> {
+        let guard = self.users.lock().unwrap();
+        Ok(guard.contains_key(username))
+    }
+
+    fn iter(&self) -> DynIter<'static, Result<String>> {
+        // Returning an iterator on something protected by a mutex = cursed.
```
Comment (gabrielmethot) [inline bin/menmosd/src/menmosd/node/test/mock.rs:70]: Wouldn't that be worthy of a TODO?

File: bin/menmosd/src/menmosd/node/test/mock.rs [hunk 6]
```diff
@@ -371,73 +366,120 @@ impl MetadataMapper for MockMetaMap {
 }
 
 #[derive(Default)]
-pub struct MockUserMap {
-    users: Mutex<HashMap<String, String>>,
+pub struct MockStorageStore {
+    m: Mutex<HashMap<String, String>>,
```
Comment (gabrielmethot) [inline bin/menmosd/src/menmosd/node/test/mock.rs:370]: Find a better variable name?

File: bin/menmosd/src/menmosd/node/test/mock.rs [hunk 7]
```diff
@@ -371,73 +366,120 @@ impl MetadataMapper for MockMetaMap {
...
 
 #[derive(Default)]
-pub struct MockIndex {
-    documents: Arc<MockDocIdMap>,
-    meta: Arc<MockMetaMap>,
-    routing: Arc<MockRoutingMap>,
-    storage: Arc<MockStorageMap>,
-    users: Arc<MockUserMap>,
+pub struct MockRoutingStore {
+    m: Mutex<HashMap<String, RoutingConfigState>>,
```
Comment (gabrielmethot) [inline bin/menmosd/src/menmosd/node/test/mock.rs:402]: Find a better variable name?

File: bin/menmosd/src/menmosd/node/test/mod.rs [hunk 8]
```diff
@@ -1,2 +1,6 @@
+//mod mock;
```
Comment (gabrielmethot) [inline bin/menmosd/src/menmosd/node/test/mod.rs:1]: Why are these commented?

File: bin/menmosd/src/menmosd/node/test/oldmock.rs [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn get_all_documents_mask(&self) -> Result<BitVec> {
+        let recycled_guard = self.recycled_ids.lock().unwrap();
+        let recycled = &*recycled_guard;
+
+        // Initialize our bitvector with 1.
```
Comment (gabrielmethot) [inline bin/menmosd/src/menmosd/node/test/oldmock.rs:92]: Why?

File: crates/menmos-interface/src/directory.rs [hunk 10]
```diff
@@ -217,30 +218,65 @@ pub trait DirectoryNode {
...
+    async fn commit(&self) -> Result<()>;
+    async fn list(&self) -> Vec<String>;
+}
+
+#[async_trait]
+pub trait QueryExecutor {
     async fn query(&self, q: &Query, username: &str) -> Result<QueryResponse>;
+    async fn query_move_requests(
+        &self,
+        q: &Query,
```
Comment (gabrielmethot) [inline crates/menmos-interface/src/directory.rs:261]: Find a better variable name?

**Review by gabrielmethot — APPROVED** [review]
You have some "Unchanged files with check annotations", all of them being unused imports. Otherwise it looks good!

File: bin/menmosd/src/menmosd/node/test/mock.rs (hunk 5 above)
Comment (dalloriam) [reply bin/menmosd/src/menmosd/node/test/mock.rs:70]: Eh not really, the method is just there so the test compiles, we don't use it so I don't really see the value.

File: bin/menmosd/src/menmosd/node/test/oldmock.rs (hunk 9 above)
Comment (dalloriam) [reply bin/menmosd/src/menmosd/node/test/oldmock.rs:92]: Its old code that I forgot to delete, but for the record its because we're creating a bitvector that contains all documents (aka a bitvector of ones).
